Detsim S1 simulation

gondiaz opened the pull request:
This PR contains the detsim functions that simulate the S1 signal from the total number of photons using the S1 LightTable. It is a single module simulate_S1.py and its corresponding test module. Also, the S1 LightTable test file is added.

This PR, along with PR #756, overrides PRs #698, #699 and #734.

mmkekic commented on invisible_cities/detsim/simulate_S1.py:
```diff
@@ -0,0 +1,194 @@
+
+    nsensors = lt.shape[-1]
+
+    if signaltype == "S1":
```
Can't you unite this two cases, there is a lot of repeated code. You'd probably need to assign `None` to `zbins` -related arrays in case of `S2` and make `get_lt_values` be able to work with that

gondiaz replied:
A `None` type in bins does not work. The best approach I found is to include in the S2 LT a z column (let's say center of the EL gap). Then, the `get_lt_value` function would contain three arguments x, y, z in both, S1 and S2 LightTables. 

In my opinion this approach is good since we can greatly reduce the extent of this function, as well as to have symmetric LTs for S1 and S2 in the sense that they contain the same column indexes. I would go for it if does not colide with #756, that I think is not the case.

mmkekic commented on invisible_cities/detsim/simulate_S1.py:
```diff
@@ -0,0 +1,194 @@
+    pes = np.random.poisson(pes)
+    return pes.T
+
+tau1 = 4.5; c1 = 0.1
```
Is there a reason why you opted to make this global?

gondiaz replied:
I am not sure if that is the best approax. We could define them as local variables inside `S1_minimize` or forget about defining the variables and put them as numbers directly.

mmkekic replied:
You should place them in `S1_minimize`. Dont forget naming conventions, functions and variables should be lowercase  only with underscores between words.  Use more descriptive names than `n, a1, a2`, as well as a more descriptive name for the function itself with an underscore in front since it should be private to this module only. Alternatively you could even place it inside `generate_S1_times` (remove capitals). Add docstring, ie that this is inverse transform sampling and `S1_minimize` (without the P random number) is inverse cdf.

mmkekic replied:
Also, `tau1, tau2` have time units (`ns`).

mmkekic commented on invisible_cities/detsim/simulate_S1.py:
```diff
@@ -0,0 +1,188 @@
+            hit position
+    Returns:
+        :pes: np.ndarray
+            photoelectrons at each PMT produced by all hits.
```
Add output shape in the description, ie (`number_of_sensors X number_of_hits`)

jacg replied:
> I am never sure how to unit test random sampling. What are the standards for this @gonzaponte , @jacg ?

Sorry, I don't have any deep wisdom to offer on this.

mmkekic commented on invisible_cities/detsim/simulate_S1.py:
```diff
@@ -0,0 +1,188 @@
+        :wfs: np.ndarray
+            waveforms with buffer_length and bin_width
+    """
+    bins = np.arange(0, buffer_length + bin_width, bin_width)
```
`bias_time` should probably be called `start_time`, `t_start` or similar.  Then you can define bins in `[t_start, t_start+buffer_length)`

mmkekic commented on invisible_cities/detsim/simulate_S1.py:
```diff
@@ -0,0 +1,188 @@
+    lt = lt.set_index(["x", "y", "z"])
+    nsensors = lt.shape[-1]
+
+    xbins=binedges_from_bincenters(np.unique(lt.index.get_level_values('x')))
```
There seems to be a numpy function that does this for you 
`histogram_bin_edges`, I believe you should call it with number of bins to be unique values +1, and specify the range you need.
However we need to decide what to do in case light table starts at `x_min> -r_active + x_pitch`, I think we should extend the lower and upper bin edge to be exactly on `r_active`. In that case we do need a custom function and `binedges_from_bincenters` can be easily adapted for this (ie adding arguments `left_edge, right_edge`). Eventually this function should then be moved to `core_functions.py`

mmkekic commented on invisible_cities/detsim/simulate_S1_test.py:
```diff
@@ -0,0 +1,95 @@
+
+
+@settings(deadline=1000)
+@given(integers(min_value=0, max_value=10000))
```
I think hypothesis is a total overkill for testing a size and a type of the output. If you do want to keep this test rename it to `test_generate_s1_time_output_type` and use some fixed size.

mmkekic commented on invisible_cities/detsim/simulate_S1.py:
```diff
@@ -0,0 +1,188 @@
+    S1_LT = create_lighttable_function(os.path.expandvars(s1_lighttable))
+
+    def create_S1_waveforms_from_hits(x, y, z, time, energy, tmin, buffer_length):
+        s1_photons     = np.random.poisson(energy / ws)
```
Maybe this doesn't belong here, lets take it out in a 
```
def num_scintillation_photons (energy, ws):
    return np.random.poisson(energy, ws) 
```
with docstrings and type annotation of course :)

mmkekic reviewed (commented):
You have to review the naming, type annotation and docstrings for all functions. 
About tests, since we are dealing with distributions I guess we should use flaky and allow some failures, but I am never sure how to unit test random sampling. What are the standards for this @gonzaponte , @jacg ?

mmkekic commented on invisible_cities/detsim/simulate_s1.py:
```diff
@@ -0,0 +1,233 @@
+    are generated.
+    Parameters:
+        :S1_pes_at_pmts: np.ndarray
+            the pes at each PMT generated by each hit
```
you are missing `hit_times`

mmkekic commented on invisible_cities/detsim/simulate_s1.py:
```diff
@@ -0,0 +1,233 @@
+
+
+def s1_waveforms_creator(s1_lighttable, ws, wf_pmt_bin_width):
+    S1_LT = create_lighttable_function(os.path.expandvars(s1_lighttable))
```
Add type hints and docstrings.

mmkekic commented on invisible_cities/detsim/simulate_s1_test.py:
```diff
@@ -0,0 +1,175 @@
+       zs = arrays(np.float, 1, elements=floats(min_value=-500, max_value=500)))
+def test_get_lt_values_s1(ICDATADIR, xs, ys, zs):
+
+    s1_lighttable = os.path.join(ICDATADIR, "NEXT_NEW.energy.S1.PmtR11410.LightTable.h5")
```
This should be a fixture (something like e15d2063ac415291c68454695791233cd22523b9 , you can probably copy the code and add S1 table as well)

mmkekic commented on invisible_cities/detsim/simulate_s1.py:
```diff
@@ -0,0 +1,233 @@
+        range = (bincenters[0], bincenters[-1])
+    else:
+        if not (range[0]<range[1]):
+            raise Exception("lower edge must be slower than higher")
```
**S**lower (:

jacg replied:
Please use a specific exception type, rather than the completely general and meaningless `Exception`!

mmkekic reviewed (commented):
Suggested placing of the function:
- `binedges_from_bincenters` should go to `core_functions`
- `create_lighttable_function` should go to new `light_tables.py` module
- `s1_waveforms_creator` should go to `detsim_waveforms.py` 

and tests should be moved to the appropriate test modules.

Then a history should be rewritten as 'added a functionality', 'added tests' for each module (ie S1 functions that stay in `simulate_s1_module`can be commited together).

Finally, a comment about the finite times of photons to reach the pmts plane should be added at the beginning of the s1 module as a reminder that this might be a problem in the future (the distribution plots can be attached to this PR)

mmkekic commented on invisible_cities/detsim/detsim_waveforms.py:
```diff
@@ -0,0 +1,73 @@
+
+def s1_waveforms_creator(s1_lighttable   : str,
+                         ws              : float,
+                         wf_pmt_bin_width: float)->Callable:
```
Add to docstrings

mmkekic commented on invisible_cities/detsim/detsim_waveforms.py:
```diff
@@ -0,0 +1,73 @@
+        :create_s1_waveforms_from_hits: function
+            see function docstring
+    """
+    S1_LT = create_lighttable_function(os.path.expandvars(s1_lighttable))
```
I know is local but I prefer it to be a lower-case name

mmkekic commented on invisible_cities/detsim/light_tables.py:
```diff
@@ -0,0 +1,65 @@
+        xindices = pd.cut(xs[sel], xbins, include_lowest=True, labels=xcenters)
+        yindices = pd.cut(ys[sel], ybins, include_lowest=True, labels=ycenters)
+        zindices = pd.cut(zs[sel], zbins, include_lowest=True, labels=zcenters)
+        indices = pd.Index(zip(xindices, yindices, zindices), name=("x", "y", "z"))
```
indentation

mmkekic commented on invisible_cities/detsim/light_tables.py:
```diff
@@ -0,0 +1,65 @@
+    if lt.get("z") is None:
+        lt.loc[:, "z"] = 1 # add fake z
+        had_z = False
+    else: had_z = True
```
put `had_z` in next line

mmkekic commented on invisible_cities/detsim/light_tables.py:
```diff
@@ -0,0 +1,65 @@
+    zbins = binedges_from_bincenters(zcenters)
+
+    if had_z: return get_lt_values
+    else:     return partial(get_lt_values, zs=np.array([1]))
```
indent

mmkekic commented on invisible_cities/detsim/simulate_s1.py:
```diff
@@ -0,0 +1,110 @@
+                           ys      : np.ndarray,
+                           zs      : np.ndarray,
+                           photons : np.ndarray,
+                           LT      : Callable  )->np.ndarray:
```
Use lower_case letters

mmkekic commented on invisible_cities/detsim/simulate_s1.py:
```diff
@@ -0,0 +1,110 @@
+    size_2 = size - size_1
+    r1 = np.random.exponential(tau1, size_1)
+    r2 = np.random.exponential(tau2, size_2)
+    return np.concatenate([r1, r2])
```
Maybe shuffle the sample before returning?

mmkekic commented on invisible_cities/detsim/simulate_s1.py:
```diff
@@ -0,0 +1,110 @@
+    It returns a list whose elements are the times at which the photoelectrons in that PMT
+    are generated.
+    Parameters:
+        :S1_pes_at_pmts: np.ndarray
```
name is not capital

mmkekic reviewed (commented):
After the cosmetics changes and some squashing the PR will be ready to be approved

mmkekic commented on invisible_cities/detsim/simulate_s1_test.py:
```diff
@@ -0,0 +1,68 @@
+    np.testing.assert_allclose(times, expected)
+
+
+def test_generate_s1_times_from_pes():
```
Can you add a test that generated times are always larger than hit times? You can set one photon per hit (even use only one pmt) and check the output times. Set largish differences in hit times.

gondiaz committed:
Add simulate_s1 module

This module contains the basic functions to simulate the S1 signal in
detsim.

gondiaz committed:
Add simulate_s1 test module

gondiaz committed:
Add light_tables module

This module contains the function that manages the s1 light table.

Add protection against indexes not in dataframe

Add function that reads light tables dataframes.

The function raises a warning if user given el_gap and/or active_r
diffear from the one used in tables files.

Call read_lt inside create_lighttable_function

Modify bin ranges

Rename reader and cosmetics

Remove unnecessary plus sign

Add explanatory comment

gondiaz committed:
Add light_tables test module

Fix reindex in tests

Modify slow tests in light_tables_test

Using hypothesis numpy arrays of element 1 is slower than direct usage
of floats strategy

gondiaz committed:
Add detsim_waveforms module

This module contains the functions that simulate the S1 signal from
hits to waveforms.

gondiaz committed:
Add detsim_waveforms test module

gondiaz committed:
Add binedges_from_bincenters function

gondiaz committed:
Add binedges_from_bincenters test

gondiaz committed:
Add S1 and S2 PMT lighttables for testing

mmkekic reviewed (approved):
This PR adds functionalities for simulation of the S1 scintillation light needed for detsim city. The code is clear and tested, good job!

mmkekic commented on invisible_cities/detsim/light_tables.py:
```diff
@@ -0,0 +1,103 @@
+    Inputs
+        fname      : string of file name
+        group_name : group name under which table is saved
+        el_gap     : optional set of different el_gap
```
review docstrings; ie should be 
```
param : type 
        explanation
```
`el_gap` and `active_r`  explanation should probably be  'optionally set a different el_gap' or similar

In a returns description `el_gap` is repeated

mmkekic commented on invisible_cities/detsim/light_tables.py:
```diff
@@ -0,0 +1,103 @@
+from .. io.dst_io  import load_dst
+
+
+def read_lt(fname, group_name, el_gap=None, active_r=None):
```
@jmalbos  suggested longer `read_lighttable` name

mmkekic commented on invisible_cities/detsim/light_tables.py:
```diff
@@ -0,0 +1,103 @@
+    zcenters = np.unique(lt.index.get_level_values('z'))
+
+    # the range for binedges is set to act_r if it is larger than the LT centers
+    range_x = (min(xcenters[0], -act_r), max(xcenters[-1], +act_r)) # centers are ordered
```
Id remove `+` in `max`

mmkekic commented on invisible_cities/detsim/light_tables.py:
```diff
@@ -0,0 +1,103 @@
+    ycenters = np.unique(lt.index.get_level_values('y'))
+    zcenters = np.unique(lt.index.get_level_values('z'))
+
+    # the range for binedges is set to act_r if it is larger than the LT centers
```
Can you expand the explanation, ie 
```
a trick to make the function work with arbitrary act_r,  extending the range if necessary.
Note that binedges_from_bincenters does not accept range smaller than centers range
```
or similar
